src/Cli/dotnet/commands/dotnet-run/VirtualProjectBuildingCommand.cs

```diff
@@ -139,86 +178,443 @@ private ProjectInstance CreateProjectInstance(
         {
             GlobalProperties = globalProperties,
         });
+
+        ProjectRootElement CreateProjectRootElement(ProjectCollection projectCollection)
+        {
+            Debug.Assert(!_directives.IsDefault && _targetFilePath is not null, $"{nameof(PrepareProjectInstance)} should have been called first.");
+
+            var projectFileFullPath = Path.ChangeExtension(EntryPointFileFullPath, ".csproj");
+            var projectFileWriter = new StringWriter();
+            WriteProjectFile(projectFileWriter, _directives, virtualProjectFile: true, targetFilePath: _targetFilePath);
+            var projectFileText = projectFileWriter.ToString();
+
+            using var reader = new StringReader(projectFileText);
+            using var xmlReader = XmlReader.Create(reader);
+            var projectRoot = ProjectRootElement.Create(xmlReader, projectCollection);
+            projectRoot.FullPath = projectFileFullPath;
+            return projectRoot;
+        }
     }
 
-    // Kept in sync with the default `dotnet new console` project file (enforced by `DotnetProjectAddTests.SameAsTemplate`).
-    private const string CommonProjectProperties = """
-            <OutputType>Exe</OutputType>
-            <TargetFramework>net10.0</TargetFramework>
-            <ImplicitUsings>enable</ImplicitUsings>
-            <Nullable>enable</Nullable>
-        """;
+    public static void WriteProjectFile(TextWriter writer, ImmutableArray<CSharpDirective> directives)
+    {
+        WriteProjectFile(writer, directives, virtualProjectFile: false, targetFilePath: null);
+    }
 
-    public static string GetNonVirtualProjectFileText()
+    private static void WriteProjectFile(TextWriter writer, ImmutableArray<CSharpDirective> directives, bool virtualProjectFile, string? targetFilePath)
     {
-        return $"""
-            <Project Sdk="Microsoft.NET.Sdk">
+        int processedDirectives = 0;
 
+        var sdkDirectives = directives.OfType<CSharpDirective.Sdk>();
+        var propertyDirectives = directives.OfType<CSharpDirective.Property>();
+        var packageDirectives = directives.OfType<CSharpDirective.Package>();
+
+        string sdkValue = "Microsoft.NET.Sdk";
+
+        if (sdkDirectives.FirstOrDefault() is { } firstSdk)
+        {
+            sdkValue = firstSdk.ToSlashDelimitedString();
+            processedDirectives++;
+        }
+
+        if (virtualProjectFile)
+        {
+            writer.WriteLine($"""
+                <Project>
```

Comment:
I've tried using XmlWriter for this, but couldn't make it write indented XML with blank lines between sections. MSBuild APIs also seem unable to do that. So I ended up building the project file as string.